optimum/onnxruntime/modeling_diffusion.py

```diff
@@ -146,13 +163,93 @@ def __init__(
             if key in all_pipeline_init_args:
                 diffusers_pipeline_args[key] = all_pipeline_init_args[key]
-        # inits diffusers pipeline specific attributes (registers modules and config)
         self.auto_model_class.__init__(self, **diffusers_pipeline_args)
 
-        # inits ort specific attributes
-        self.shared_attributes_init(
-            model=unet_session, use_io_binding=use_io_binding, model_save_dir=model_save_dir, **kwargs
-        )
+        # Forced on every class inheriting from OptimizedModel
+        self.preprocessors = kwargs.pop("preprocessors", [])
+
+    @property
+    def components(self) -> Dict[str, Any]:
+        components = {
+            "vae": self.vae,
+            "unet": self.unet,
+            "text_encoder": self.text_encoder,
+            "text_encoder_2": self.text_encoder_2,
+            "safety_checker": self.safety_checker,
+            "image_encoder": self.image_encoder,
+        }
+        components = {k: v for k, v in components.items() if v is not None}
+        return components
+
+    def _validate_same_attribute_value_across_components(self, attribute: str):
+        # The idea is that these attributes make sense for the pipeline as a whole only when they are the same across
+        # all components, so we do support these attributes but also allow the user to experiment with undefined behavior
+        # like having heterogeneous devices or io bindings across components.
+        attribute_values = {
+            name: getattr(component, attribute)
+            for name, component in self.components.items()
+            if hasattr(component, attribute)
+        }
+
+        if len(attribute_values) == 0:
+            raise ValueError(f"Attribute {attribute} is not defined for any component.")
+        # make sure there is exactly one value for the attribute (bypass unhashable types)
+        elif len(set(map(str, attribute_values.values()))) > 1:
+            raise ValueError(f"Attribute {attribute} is not the same across components: {attribute_values}.")
+
+        return next(iter(attribute_values.values()))
+
+    @property
+    def device(self) -> torch.device:
+        return self._validate_same_attribute_value_across_components("device")
+
+    @property
+    def dtype(self) -> torch.dtype:
+        return self._validate_same_attribute_value_across_components("dtype")
+
+    @property
+    def providers(self) -> Tuple[str]:
+        return self._validate_same_attribute_value_across_components("providers")
+
+    @property
+    def provider(self) -> str:
+        return self._validate_same_attribute_value_across_components("provider")
+
+    @property
+    def providers_options(self) -> Dict[str, Dict[str, Any]]:
+        return self._validate_same_attribute_value_across_components("providers_options")
+
+    @property
+    def provider_options(self) -> Dict[str, Any]:
+        return self._validate_same_attribute_value_across_components("provider_options")
```

Comment:
It is not necessary to validate same value across components.

I think it is feasible to use different provider and different provider options for components. For example, we can run text_encoder with CPU, and unet with CUDA provider. Or we want to enable cuda graph in one component but not the other in provider option.

May add some comments and loose the constraint later.

Reply:
there's a comment in `_validate_same_attribute_value_across_components` definition explaining the reasoning behind these checks, which is exactly what you said. Pipeline attributes can be accessed but they only make sense when they're consistent, for now this is my proposition for multi model parts pipelines, an alternative would be to return that of the main component (unet/transformer) or not supporting these attributes at all for the main pipeline.